cmd/nerdctl/login_linux_test.go

```diff
@@ -57,269 +52,485 @@ func (ag *Client) WithHostsDir(hostDirs string) *Client {
+	}
 	return ag
 }
 
+func (ag *Client) WithConfigPath(value string) *Client {
+	ag.configPath = value
+	return ag
+}
+
+func (ag *Client) GetConfigPath() string {
+	return ag.configPath
+}
+
 func (ag *Client) Run(base *testutil.Base, host string) *testutil.Cmd {
 	if ag.configPath == "" {
 		ag.configPath, _ = os.MkdirTemp(base.T.TempDir(), "docker-config")
 	}
-	args := append([]string{"--debug-full", "login"}, ag.args...)
+	args := []string{"login"}
+	if base.Target == "nerdctl" {
+		args = append(args, "--debug-full")
+	}
+	args = append(args, ag.args...)
 	icmdCmd := icmd.Command(base.Binary, append(base.Args, append(args, host)...)...)
 	icmdCmd.Env = append(base.Env, "HOME="+os.Getenv("HOME"), "DOCKER_CONFIG="+ag.configPath)
+
 	return &testutil.Cmd{
 		Cmd:  icmdCmd,
 		Base: base,
 	}
 }
 
-func TestLogin(t *testing.T) {
+func TestLoginPersistence(t *testing.T) {
+	base := testutil.NewBase(t)
+	t.Parallel()
+
+	// Retrieve from the store
+	testCases := []struct {
+		auth string
+	}{
+		{
+			"basic",
+		},
+		{
+			"token",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("Server %s", tc.auth), func(t *testing.T) {
+			t.Parallel()
+
+			username := testregistry.SafeRandomString(30) + "∞"
+			password := testregistry.SafeRandomString(30) + ":∞"
+
+			// Add the requested authentication
+			var auth testregistry.Auth
+			var dependentCleanup func(error)
+
+			auth = &testregistry.NoAuth{}
+			if tc.auth == "basic" {
+				auth = &testregistry.BasicAuth{
+					Username: username,
+					Password: password,
+				}
+			} else if tc.auth == "token" {
+				authCa := testca.New(base.T)
+				as := testregistry.NewAuthServer(base, authCa, 0, username, password, false)
+				auth = &testregistry.TokenAuth{
+					Address:  as.Scheme + "://" + net.JoinHostPort(as.IP.String(), strconv.Itoa(as.Port)),
+					CertPath: as.CertPath,
+				}
+				dependentCleanup = as.Cleanup
+			}
+
+			// Start the registry with the requested options
+			reg := testregistry.NewRegistry(base, nil, 0, auth, dependentCleanup)
+
+			// Register registry cleanup
+			t.Cleanup(func() {
+				reg.Cleanup(nil)
+			})
+
+			// First, login successfully
+			c := (&Client{}).
+				WithCredentials(username, password)
+
+			c.Run(base, fmt.Sprintf("localhost:%d", reg.Port)).
+				AssertOK()
+
+			// Now, log in successfully without passing any explicit credentials
+			nc := (&Client{}).
+				WithConfigPath(c.GetConfigPath())
+			nc.Run(base, fmt.Sprintf("localhost:%d", reg.Port)).
+				AssertOK()
+
+			// Now fail while using invalid credentials
+			nc.WithCredentials("invalid", "invalid").
+				Run(base, fmt.Sprintf("localhost:%d", reg.Port)).
+				AssertFail()
+
+			// And login again without, reverting to the last saved good state
+			nc = (&Client{}).
+				WithConfigPath(c.GetConfigPath())
+
+			nc.Run(base, fmt.Sprintf("localhost:%d", reg.Port)).
+				AssertOK()
+		})
+	}
+}
+
+/*
+func TestAgainstNoAuth(t *testing.T) {
+	base := testutil.NewBase(t)
```

Comment:
is this WIP?

Reply:
Follow-up PR to this will introduce no-auth registry testing once I fix the login logic on top of this PR.

I can remove this for now if this is preferable.

LMK your preference.